devicedefender/include/aws/iotdevicedefender/DeviceDefender.h

```diff
@@ -77,6 +77,232 @@ namespace Aws
+             * other device defender payload data with the metric name of "metricName".
+             * @param metricName The key name for the data.
+             * @param metricFunc The function that is called to get the number data.
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *          Will return AWS_OP_ERR if the number cannot be registered or the data passed
+             *          ("metricName" or "metricFunc") is incorrect.
+             */
+            int RegisterCustomMetricNumber(
+                aws_byte_cursor metricName,
+                std::function<int(int64_t *, void *)> *metricFunc);
+
+            /**
+             * Registers a custom metric number list function to the Device Defender result. Will call the "metricFunc"
+             * function that is passed in each time a report is generated so it's data can be passed along with the
+             * other device defender payload data with the metric name of "metricName".
+             *
+             * Note: You do not need to initialize the aws_array_list! It is already initialized when passed to
+             * "metricFunc".
+             * @param metricName The key name for the data.
+             * @param metricFunc The function that is called to get the number list data.
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *          Will return AWS_OP_ERR if the number list cannot be registered or the data passed
+             *          ("metricName" or "metricFunc") is incorrect.
+             */
+            int RegisterCustomMetricNumberList(
+                aws_byte_cursor metricName,
+                aws_iotdevice_defender_get_number_list_fn *metricFunc);
+
+            /**
+             * Registers a custom metric number list function to the Device Defender result. Will call the "metricFunc"
+             * function that is passed in each time a report is generated so it's data can be passed along with the
+             * other device defender payload data with the metric name of "metricName".
+             *
+             * Note: You do not need to initialize the aws_array_list! It is already initialized when passed to
+             * "metricFunc".
+             * @param metricName The key name for the data.
+             * @param metricFunc The function that is called to get the number list data.
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *          Will return AWS_OP_ERR if the number list cannot be registered or the data passed
+             *          ("metricName" or "metricFunc") is incorrect.
+             */
+            int RegisterCustomMetricNumberList(
+                aws_byte_cursor metricName,
+                std::function<int(aws_array_list *, void *)> *metricFunc);
```

Comment:
I'd even be tempted to let the callback *return* a std::vector instead of modify one in-place

but I'm not 100% sure what the compiler would end up doing with that, if it would end up another needless copy or not. I don't understand [copy-ellision](https://en.cppreference.com/w/cpp/language/copy_elision) and return-value-optimization well enough

but in other languages I *would* let the callback just return the vector by value, because I know that wouldn't be an extra copy in other languages